pghoard/common.py

```diff
@@ -100,6 +104,70 @@ class BaseBackupMode(StrEnum):
     pipe = "pipe"
 
 
+class ProgressData(BaseModel):
+    current_progress: float = 0
+    last_updated_time: float = 0
+
+    @property
+    def age(self) -> float:
+        return time.time() - self.last_updated_time
+
+    def update(self, current_progress: float) -> None:
+        self.current_progress = current_progress
+        self.last_updated_time = time.time()
+
+
+def atomic_write(file_path: str, data: str, temp_dir: Optional[str] = None):
+    temp_dir = temp_dir or os.path.dirname(file_path)
+    temp_file = None
+    try:
+        with tempfile.NamedTemporaryFile("w", delete=False, dir=temp_dir) as temp_file:
+            temp_file.write(data)
+            temp_path = temp_file.name
+        os.rename(temp_path, file_path)
+    except Exception as ex:  # pylint: disable=broad-except
+        LOG.exception("Failed to write file atomically: %r", ex)
+        if temp_file and os.path.exists(temp_file.name):
```

Comment:
It's better to just try and remove it and suppress the exception, as there is a race condition here between where you check the file exists and call remove. It's possible for the first to succeed, then something remove the file before the unlink call is made.